pkg/streamingpromql/aggregations/max.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package aggregations
+
+import (
+	"math"
+
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type MaxAggregationFunction struct {
+	// Sum, presence, and histograms for each step.
+	floatValues  []float64
+	floatPresent []bool
+}
+
+func (g *MaxAggregationFunction) AccumulateSeries(data types.InstantVectorSeriesData, steps int, start int64, interval int64, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, _ functions.EmitAnnotationFunc) error {
+	var err error
+	if len(data.Floats) > 0 && g.floatValues == nil {
+		// First series with float values for this group, populate it.
+		g.floatValues, err = types.Float64SlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.floatPresent, err = types.BoolSlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+		g.floatValues = g.floatValues[:steps]
+		g.floatPresent = g.floatPresent[:steps]
+	}
+
+	for _, p := range data.Floats {
+		if math.IsNaN(p.F) {
+			continue
+		}
+		idx := (p.T - start) / interval
+		if !g.floatPresent[idx] || g.floatPresent[idx] && p.F > g.floatValues[idx] {
+			g.floatValues[idx] = p.F
+			g.floatPresent[idx] = true
+		}
```

Comment:
[nit] Have you considered extracting this logic (and the equivalent logic below for histograms) to a single function? I'm imagining we could then call something like `accumulatePoint(p.T, p.F)` for floats and `accumulatePoint(p.T, 0)` for histograms.